### PR TITLE
ACS-6304 Add retry mechanism to completeRecordsCantBeUpdated test

### PR DESCRIPTION
**Pull request by mikolajbrzezinski**
The `completeRecordsCantBeUpdated` test fails intermittently in both community-repo and enterprise-repo. We can mitigate this issue by rerunning the failed test using an implementation of `RetryAnalyzerCount` abstract class.

### CODE REVIEW

File: amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * GNU Lesser General Public License for more details.
+ * -
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+package org.alfresco.rest.rm.community.utils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (dsibilio) [inline amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java:30]: Unused imports

File: amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        if (status == 1)
+        {
+            resultName = "SUCCESS";
+        }
+        if (status == 2)
+        {
+            resultName = "FAILURE";
+        }
+        return resultName;
+    }
```
Comment (dsibilio) [inline amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java:51]: Why is this here? It doesn't seem to be used anywhere.

File: amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestResult;
+
+public class RetryAnalyzer implements IRetryAnalyzer
+{
+    private int retryCount;
+    private int maxRetryCount = 3;
```
Comment (dsibilio) [inline amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java:37]: This could be a constant.

File: amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            else
+            {
+                result.setStatus(ITestResult.FAILURE);
+            }
+        }
+        else
+        {
+            result.setStatus(ITestResult.SUCCESS);
+        }
+        return false;
```
Comment (dsibilio) [inline amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/RetryAnalyzer.java:73]: I'm puzzled by this logic:
1. why are we manually setting the test results? The framework should handle that automatically depending on the retry's outcome.
2. why are we checking that the test failed? The `IRetryAnalyzer#retry` method should only be invoked by the framework if the test failed.

**Review by dsibilio — CHANGES_REQUESTED** [review]

**Comment by dsibilio** [comment]
While I appreciate you getting your hands dirty with a custom implementation, I've found a default TestNG `IRetryAnalyzer` abstract implementation called [`RetryAnalyzerCount`](https://javadoc.io/doc/org.testng/testng/6.8.8/org/testng/util/RetryAnalyzerCount.html) that only requires overriding the max retry count and the `retryMethod`. It would be better to re-use that instead.

File: amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/AlfrescoRetryAnalyzer.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+import org.testng.ITestResult;
+import org.testng.util.RetryAnalyzerCount;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class AlfrescoRetryAnalyzer extends RetryAnalyzerCount
+{
+    private int count = 4;
+
+    public boolean retry(ITestResult result)
```
Comment (dsibilio) [inline amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/utils/AlfrescoRetryAnalyzer.java:37]: If we're overriding methods we should always make this clear via the **@Override** annotation. This is not only useful to better understand at first glance what is going on, but also provides compilation time feedback in case the parent signature changes in the future. This class will silently break and no one would ever know if someone changes the signature of the method we're overriding since we're not using the appropriate marker annotation.

**Review by aonych — APPROVED** [review]

**Review by dsibilio — APPROVED** [review]

File: amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/records/UpdateRecordsTests.java [hunk 6]
```diff
@@ -95,7 +96,7 @@ public void preconditionUpdateRecordsTests()
     {
         rootCategory = createRootCategory(getRandomName("CATEGORY NAME"));
         unfiledRecordFolder = createUnfiledContainerChild(UNFILED_RECORDS_CONTAINER_ALIAS,
-                getRandomName("UnfiledRecordFolder"), UNFILED_RECORD_FOLDER_TYPE);
+            getRandomName("UnfiledRecordFolder"), UNFILED_RECORD_FOLDER_TYPE);
```
Comment (dsibilio) [inline amps/ags/rm-automation/rm-automation-community-rest-api/src/test/java/org/alfresco/rest/rm/community/records/UpdateRecordsTests.java:99]: We can revert this, unnecessary change / not in line with the rest of the formatting for the method.

**Comment by dsibilio** [comment]
Changed the title from **ACS-6304 Fix completeRecordsCantBeUpdated intermittent failure** to **ACS-6304 Add retry mechanism to completeRecordsCantBeUpdated test**.
For the future, let's try to be more descriptive since the PR title should reflect the squashed commit on master. Terms like "Fix" are generally okay but if there's an easy way of adding more information with very minimal additional wording then it's just a hindrance as it makes the commit log less meaningful.
